dwave/optimization/src/nodes/manipulation.cpp

```diff
@@ -18,6 +18,163 @@
+    return data_ptr<ArrayNodeStateData>(state)->buff();
+}
+
+std::span<const Update> ConcatenateNode::diff(const State& state) const {
+   return data_ptr<ArrayNodeStateData>(state)->diff();
+}
+
+void ConcatenateNode::initialize_state(State& state) const {
+    int index = topological_index();
+    assert(index >= 0 && "must be topologically sorted");
+    assert(static_cast<int>(state.size()) > index && "unexpected state length");
+    assert(state[index] == nullptr && "already initialized state");
+
+    std::vector<double> values;
+    values.reserve(size());
+
+    // Prefix is defined over dimensions 0..axis-1
+    std::vector<std::vector<ssize_t>> prefix_dims(axis_);
+    for (ssize_t dim = 0, stop = prefix_dims.size(); dim < stop; ++dim) {
+        for (ssize_t i = 0, stop = shape()[dim]; i < stop; ++i) {
+            prefix_dims[dim].push_back(i);
+        }
+    }
+
+    // Suffix is defined over dimensions axis+1..ndim
+    std::vector<std::vector<ssize_t>> suffix_dims(ndim() - axis_ - 1);
+    for (ssize_t dim = 0, stop = suffix_dims.size(); dim < stop; ++dim) {
+        for (ssize_t i = 0, stop = shape()[axis_ + 1 + dim]; i < stop; ++i) {
+            suffix_dims[dim].push_back(i);
+        }
+    }
+
+    auto prefix_prod = cartesian_product(prefix_dims);
+    auto suffix_prod = cartesian_product(suffix_dims);
+
+    std::vector<ssize_t> indices;
+    for (auto prefix : prefix_prod) {
+        for (ssize_t arr_i = 0, stop = array_ptrs_.size(); arr_i < stop; ++arr_i) {
+            for (ssize_t arr_axis_i = 0, stop = array_ptrs_[arr_i]->shape()[axis_]; arr_axis_i < stop; ++arr_axis_i) {
+                for (auto suffix : suffix_prod) {
+                    indices.insert(indices.begin(), prefix.begin(), prefix.end());
+                    indices.insert(indices.begin() + prefix_dims.size(), arr_axis_i);
+                    indices.insert(indices.begin() + prefix_dims.size() + 1, suffix.begin(), suffix.end());
+
+                    ssize_t idx = ravel_multi_index(array_ptrs_[arr_i]->strides(), indices);
+                    values.emplace_back(array_ptrs_[arr_i]->view(state)[idx]);
+                    indices.clear();
+                }
+            }
+        }
+    }
+
+    state[index] = std::make_unique<ArrayNodeStateData>(std::move(values));
+}
+
+void ConcatenateNode::commit(State& state) const {
+    data_ptr<ArrayNodeStateData>(state)->commit();
+}
+
+void ConcatenateNode::revert(State& state) const {
+    data_ptr<ArrayNodeStateData>(state)->revert();
+}
+
+void ConcatenateNode::propagate(State& state) const {
+    auto ptr = data_ptr<ArrayNodeStateData>(state);
+    ssize_t offset = 0;
+
+    for (auto it = array_ptrs_.begin(), stop = array_ptrs_.end(); it != stop; ++it) {
+        if (std::distance(array_ptrs_.begin(), it) > 0) {
+            offset += (*std::prev(it))->shape()[axis_];
+        }
+
+        for (auto diff : (*it)->diff(state)) {
+            auto indices = unravel_index((*it)->strides(), diff.index);
```

Comment:
I mentioned in our chat that this should be correct, but looking at it again, I'm questioning myself. The reason is that I believe the update index (`diff.index`) is actually a linear index assuming "standard" strides of the originating node (we should really document this better @arcondello). Thus, using the `strides()` of the predecessor may cause this computed multi-index to be wrong if the predecessor has non-standard strides.

To me this adds more weight that we should think about requiring array nodes to provide a `standardized_strides()` or something that returns the strides based only the shape. Otherwise, I think you'll have to compute the "standard" strides for each predecessor to be able to use `unravel_index()` here.

Reply:
I see, do you have an example of non standard strides?

Reply:
I think @wbernoudy means contiguous strides rather than standard. I.e. what would the strides be for a contiguous array.

You can make a non-contiguous array easily with a `BasicIndexingNode`. Something like
```c++
graph.emplace_node<BasicIndexingNode>(x_ptr, Slice(std::nullopt, std::nullopt, 2))
```
this is the same as
```python
x[::2]
```
in Python.

Reply:
Yes, thanks, I should have been saying non-contiguous instead of non-standard.

Reply:
Should be fixed now with the use of random access ArrayIterator :)